Export Sector Status (#1220)
filecoin-project/FIPs: FIPS/fip-0112.md

---
fip: "0112"
title: Export Sector Status to FEVM 
author: "@Kubuxu @lordshashank @ZenGround0"
discussions-to: https://github.com/filecoin-project/FIPs/discussions/1108
status: Draft
type: Technical (Core)
category: Core
created: 2025-11-26
spec-sections: 
  - System Actors (miner actor) 
---

<!--You can leave these HTML comments in your merged FIP and delete the visible duplicate text guides, they will not appear and may be helpful to refer to if you edit it again. This is the suggested template for new FIPs. Note that a FIP number will be assigned by an editor. When opening a pull request to submit your FIP, please use an abbreviated title in the filename, `fip-draft_title_abbrev.md`. The title should be 44 characters or less.-->

# FIP-0112: Export Sector Status to FEVM

> ## Simple Summary

Reviewer: Abstract and Simple Summary are identical. Maybe just get AI to make you a single para Simple Summary.

<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the FIP.-->
We propose exporting a new method `validate_sector_status` from the builtin miner actor for creating user space FEVM smart contracts that can check the status of filecoin sectors, either live or dead, on chain.  An additional exported method `generate_sector_status_info` can be used in an offchain call to generate inputs to `validate_sector_status`.

> Combined with the recent changes to ProveCommitSectors3 in FIP-0109 allowing smart contract notifications upon sector commitent, this FIP establishes a minimally useful interface for a rich set of programmable storage applications.

Reviewer: ```suggestion
Combined with the recent changes to ProveCommitSectors3 in FIP-0109 allowing smart contract notifications upon sector commitment, this FIP establishes a minimally useful interface for a rich set of programmable storage applications.
```

    pub partition: i64,
}
```

### validate_sector_status

This method takes in a sector number, generic serialized bytes encoding a sector location and the status of the sector, either `Live` or `Dead`.  It returns true if the correct sector status is passed.

This method uses the sector location to look up the sector's partition and determine if the `Sectors` bitfield includes the specified sector number.  If it is validating a declared `Live` sector it checks that the `Terminated` bitfield does not include the sector number.  If the sector is not found at the provided index the method errors.

When validating a `Dead` sector with a partition and deadline index > 0 this method performs the same lookup and verifies that the sector number is present in the `Terminated` bitfield.  If the sector is not found at the provided index the method errors.  If the the partition and deadline index are NO_PARTITION and NO_DEADLINE then the method verifies that the sector number is not assigned to any member of the miner actor's sectors AMT.

Input type:
```rust
pub struct ValidateSectorStatusParams {
    pub sector_number: SectorNumber,
    pub status: SectorStatusCode,

>     pub aux_data: Vec<u8>,

Author: Responded below so resolving this reference.

SectorNumber
```

Return type:
```rust
ChainEpoch
```

## Design Rationale
<!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->

We explicitly have chosen to make the main status interface a validating interface versus a data retrieval interface.  This design puts the burden of index search/storage and making data available on off chain callers since the caller to the validate method must already have looked up the answer to the query they are validating.  This forces designs which are ignorant of miner actor bookkeeping decisions allowing freedom to evolve miner actor design without breaking external interfaces.  Today there are calls, notably `compact_partitions`, which can change a sector's partition index.  So this design helpfully steers smart contract developers away from relying on unstable system internals.

> The `aux_data` field in the input to `ValidateSectorStatusInfoParams` further allows us to evolve the method with changes to the builtin miner actor.

Author: The reason to include this now instead of relying on the `ProveCommit` `ProveCommitSectors2` `ProveCommitSectors3` development approach by default is for smart contracts to be able to include their own future proofing safeguards.  If calling contracts allow users to pass bytes in today then the smart contract calling on sector status doesn't need to upgrade to deal with builtin actors upgrades.

However I think future proofing concerns are fairly esoteric at the current rate of development in the ecosystem.  So I could be convinced to leave this off.

We've included the convenient `generate_sector_status_info` method to provide an easy canonical way to generate sector location indexes, though we expect pre-existing offchain indexes to be considered as alternative sources of sector location when adding support for this change into fullnodes.

We forwent a similar validating interface pattern with `get_sector_expiration`.  This would look like a method that takes in the correct expiration and returns True or False depending on whether the input expiration is correct.  This is done for expediency and convenience.  The thinking is that this method only ever provides a useful hint anyway as it cannot provide any useful information when handling early terminations.

> We considered including the most current guess of a sectors expiration from `get_sector_expiration`, which could be accomplished by providing the sector location and traversing the partition's expiration queue. This would effectively expose a signal of a sector's faultiness which runs counter to our design of `validate_sector_status`. If we want to expose this information we should be consistent and create a new `Fault` sector status value from the enum.

Reviewer: ```suggestion
We considered including the most current guess of a sector's expiration from `get_sector_expiration`, which could be accomplished by providing the sector location and traversing the partition's expiration queue. This would effectively expose a signal of a sector's faultiness which runs counter to our design of `validate_sector_status`. If we want to expose this information we should be consistent and create a new `Fault` sector status value from the enum.
```

The interface with the filecoin sector storage system was chosen to support the authors' concept of a minimally viable FEVM sector storage service.  Periodically checking liveness of a sector onchain is a simple way to create a storage system verified in user space.  And it is not too onerous on participating storage providers.

There has been some discussion of exposing finer grained sector status details, i.e. the unproven and faulty states. This would add complexity to the interface for limited benefit.  While exposing Filecoin internal states might be useful for power users wishing to essentially extend the core filecoin protocol in specific ways, i.e. add additional fault fees for certain sectors, this FIP is focused on user applications that can be served by a simpler model of filecoin storage.  By way of internal incentives and proofs filecoin sectors are considered live until they are terminated or expired.  Removing internal hints such as faults from the user's view creates a model that is easier to reason about and design against.

> There has also been discussion on not inlcuding sector expiration information.  Expiration information has limitations in usefullness given that early termination can happen at any time and invalidate expiration.  However the cost is low and the interface is simple especially with a decoupled method specifically for retrieving expirations.

Reviewer: ```suggestion
There has also been discussion on not including sector expiration information.  Expiration information has limitations in usefulness given that early termination can happen at any time and invalidate expiration.  However the cost is low and the interface is simple especially with a decoupled method specifically for retrieving expirations.
```

## Implementation
<!--The implementations must be completed before any core FIP is given status "Final", but it need not be completed before the FIP is accepted. While there is merit to the approach of reaching consensus on the specification and rationale before writing code, the principle of "rough consensus and running code" is still useful when it comes to resolving many discussions of API details.-->
WIP builtin actors PR here: https://github.com/filecoin-project/builtin-actors/pull/1707

## TODO

> <!--A section that lists any unresolved issues or tasks that are part of the FIP proposal. Examples of these include performing benchmarking to know gas fees, validate claims made in the FIP once the final implementation is ready, etc. A FIP can only move to a “Last Call” status once all these items have been resolved.-->

Reviewer: is there anything we can say at this point about gas for these?
